File: backend/app/schemas/project_copy.py
from pydantic import BaseModel, Field, model_validator

from app.models.recipe_generation_mode import RecipeGenerationMode


class ProjectCopyRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    participants: int = Field(gt=0)
    days: int = Field(gt=0)
    start_date: str | None = Field(default=None, max_length=20)
    first_meal: str | None = Field(default=None, max_length=20)
    last_meal: str | None = Field(default=None, max_length=20)
    recipe_generation_mode: RecipeGenerationMode = RecipeGenerationMode.CLUB_ONLY

    @model_validator(mode="after")
    def validate_meal_boundaries(self) -> "ProjectCopyRequest":
        if (self.first_meal is None) != (self.last_meal is None):
            raise ValueError("First and last meal must be provided together")
        return self


class ProjectCopyResponse(BaseModel):
    project_id: int
    meal_plan_id: str
    copied_slot_count: int = Field(ge=0)
    copied_assignment_count: int = Field(ge=0)
    skipped_assignment_count: int = Field(ge=0)
    warnings: list[str]

File: frontend/src/features/project/model/projectCopy.ts
import type { CreateProjectRequest, Project, ProjectCopyResponse } from "../api/projectApi";

export function buildProjectCopyDefaults(source: Project): CreateProjectRequest {
  return {
    name: `${source.name} — копия`,
    participants: source.participants,
    days: source.days,
    start_date: source.start_date ?? undefined,
    first_meal: source.first_meal ?? undefined,
    last_meal: source.last_meal ?? undefined,
    recipe_generation_mode: source.recipe_generation_mode,
  };
}

export function projectCopySummary(result: ProjectCopyResponse): string {
  const skipped = result.skipped_assignment_count
    ? ` Пропущено назначений: ${result.skipped_assignment_count}.`
    : "";
  return `Скопировано слотов: ${result.copied_slot_count}, назначений: ${result.copied_assignment_count}.${skipped}`;
}
